### PR TITLE
Issue 408 - add file line number to search suggestion

### PR DESCRIPTION
**Pull request by maxhallinan**
screenshot from 2019-02-15 07-17-48

When a user searches for `filename:123` and there is one suggested file, then the suggestion links to that file at line 123. More discussion about this feature is found on issue #408. There is no change to the behavior of the search page except that a search for `filename:123` is now implicitly a search for `filename`. `:123` is removed from the query by the frontend.

### CODE REVIEW

File: web/src/tree/Tree.tsx [hunk 1]
```diff
@@ -272,7 +272,10 @@ export class Tree extends React.PureComponent<Props, State> {
                     // Strip the ?suggestion query param. Handle both when going from ancestor -> child and child -> ancestor.
                     if (queryParams.has('suggestion')) {
                         queryParams.delete('suggestion')
-                        this.props.history.replace({ search: queryParams.toString() })
+                        this.props.history.replace({
+                            ...this.props.location,
+                            search: queryParams.toString(),
+                        })
```
Comment (maxhallinan) [inline web/src/tree/Tree.tsx]: The intention is just to remove `?suggestion` but this was also clobbering the rest of the location state. At least the line number hash should be preserved. Let me know if a more explicit approach is better here. My thinking is that merging all of `this.props.location` into this update will save someone some debugging time later. It immediately obvious to me where `#L123` was being removed.

File: web/src/search/input/QueryInput.tsx [hunk 2]
```diff
@@ -120,25 +123,41 @@ export class QueryInput extends React.Component<Props, State> {
...
+                        this.props.onChange(queryWithoutLineNumber)
+                    }),
                     distinctUntilChanged(),
                     debounceTime(200),
                     switchMap(query => {
                         if (query.length < QueryInput.SUGGESTIONS_QUERY_MIN_LENGTH) {
                             return [{ suggestions: [], selectedSuggestion: -1, loading: false }]
                         }
+                        // Match line number in this form: <query>:123
+                        const [, lineDigits = ''] = this.state.queryWithLineNumber.match(/:(\d+)$/) || []
```
Comment (felixfbecker) [inline web/src/search/input/QueryInput.tsx]: Wouldn't this match something like `foo bar max:1000` too?

**Comment by sqs** [comment]
This is a great feature addition! From a product POV a query with file:lineno should also show that line on the search results page (not just in query suggestions). It is ok to merge this PR without that (ie with it only working in query suggestions) if it’s feature flagged (just using a localStorage feature flag is fine).

File: web/src/search/input/QueryInput.tsx (hunk 2 above)
Comment (maxhallinan) [reply web/src/search/input/QueryInput.tsx]: Yes, good point!

**Comment by maxhallinan** [comment]
@felixfbecker I agree. There was some more discussion on #408 after I opened this PR. What do you think about the approach on #2333 for handling line numbers in Search queries? If that approach is better, maybe the same thing can be done for SearchSuggestions.

**Comment by maxhallinan** [comment]
Closing this because it's clearly not the right approach. Thanks for the review @felixfbecker

**Pull request closed without merging**